src/data/equiv/basic.lean

```diff
@@ -751,6 +751,55 @@ def sum_compl {α : Type*} (p : α → Prop) [decidable_pred p] :
   (a : α) (h : ¬ p a) :
   (sum_compl p).symm a = sum.inr ⟨a, h⟩ := dif_neg h
 
+open equiv
+
+variables {ε : Type*} {p : ε → Prop} [decidable_pred p]
+variables (ep ep' : perm {a // p a}) (en en' : perm {a // ¬ p a})
+
+/-- Combining permutations on `ε` that permute only inside or outside the subtype
+split induced by `p : ε → Prop` constructs a permutation on `ε`. -/
+def perm.subtype_congr : equiv.perm ε :=
+perm_congr (sum_compl p) (sum_congr ep en)
+
+lemma perm.subtype_congr.apply (a : ε) :
+  ep.subtype_congr en a = dite (p a) (λ h, ep ⟨a, h⟩) (λ h, en ⟨a, h⟩) :=
+by { by_cases h : p a; simp [perm.subtype_congr, h] }
+
+@[simp] lemma perm.subtype_congr.left_apply {a : ε} (h : p a) :
+  ep.subtype_congr en a = ep ⟨a, h⟩ :=
+by simp [perm.subtype_congr.apply, h]
+
+@[simp] lemma perm.subtype_congr.right_apply {a : ε} (h : ¬ p a) :
+  ep.subtype_congr en a = en ⟨a, h⟩ :=
+by simp [perm.subtype_congr.apply, h]
```

Comment:
Are conditional simp lemmas ok?

Reply:
`dif_pos` and `dif_neg` are conditional simp lemmas, and these ones are basically overpowered versions of `dif_pos` and `dif_neg`.

Reply:
Looks like `function.update_noteq` is too - I'm convinced :)